protokube/pkg/protokube/hetzner_volume.go

```diff
@@ -0,0 +1,119 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package protokube
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"os"
+	"strconv"
+
+	"github.com/hetznercloud/hcloud-go/hcloud"
+	"github.com/hetznercloud/hcloud-go/hcloud/metadata"
+	"k8s.io/klog/v2"
+	"k8s.io/kops/protokube/pkg/gossip"
+	gossiphetzner "k8s.io/kops/protokube/pkg/gossip/hetzner"
+	"k8s.io/kops/upup/pkg/fi/cloudup/hetzner"
+)
+
+// HetznerVolumes defines the Hetzner Cloud volume implementation.
+type HetznerVolumes struct {
+	hcloudClient *hcloud.Client
+	server       *hcloud.Server
+}
+
+func (h HetznerVolumes) AttachVolume(volume *Volume) error {
+	//TODO implement me
+	panic("implement me")
+}
+
+func (h HetznerVolumes) FindVolumes() ([]*Volume, error) {
+	//TODO implement me
+	panic("implement me")
+}
+
+func (h HetznerVolumes) FindMountedVolume(volume *Volume) (device string, err error) {
+	//TODO implement me
+	panic("implement me")
+}
+
+var _ Volumes = &HetznerVolumes{}
+
+// NewHetznerVolumes returns a new Hetzner Cloud volume provider.
+func NewHetznerVolumes() (*HetznerVolumes, error) {
+	serverID, err := metadata.NewClient().InstanceID()
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve server id: %s", err)
+	}
+	klog.V(4).Infof("Found ID of the running server: %d", serverID)
+
+	hcloudToken := os.Getenv("HCLOUD_TOKEN")
+	if hcloudToken == "" {
+		return nil, fmt.Errorf("%s is required", "HCLOUD_TOKEN")
+	}
+	opts := []hcloud.ClientOption{
+		hcloud.WithToken(hcloudToken),
+	}
+	hcloudClient := hcloud.NewClient(opts...)
+
+	server, _, err := hcloudClient.Server.GetByID(context.TODO(), serverID)
+	if err != nil || server == nil {
+		return nil, fmt.Errorf("failed to get info for the running server: %s", err)
+	}
+
+	klog.V(4).Infof("Found name of the running server: %q", server.Name)
+	if server.Datacenter != nil && server.Datacenter.Location != nil {
+		klog.V(4).Infof("Found location of the running server: %q", server.Datacenter.Location.Name)
+	} else {
+		return nil, fmt.Errorf("failed to find location of the running server")
+	}
+	if len(server.PrivateNet) > 0 {
```

Comment:
Not a blocker, but looked like you might be able to get this from private-networks in metadata?  https://docs.hetzner.cloud/#server-metadata

Although I suppose you're getting the server anyway, though if that's only to get the cluster label I thought you had a PR to pass that in as a parameter...

Reply:
Server metadata is returned mostly unparsed by the lib so would have meant a little more work. We need to list all peers anyway later, so using the cloud API is still needed. Added a TODO to do it later.